nemo/collections/asr/parts/mixins/streaming.py

```diff
@@ -0,0 +1,79 @@
+# Copyright (c) 2022, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import ABC, abstractmethod
+
+import torch
+
+
+class StreamingEncoder(ABC):
+    @abstractmethod
+    def setup_streaming_params(
+        self, init_chunk_size=None, init_shift_size=None, chunk_size=None, shift_size=None, cache_drop_size=None
+    ):
+        pass
+
+    @abstractmethod
+    def get_initial_cache_state(self, batch_size, dtype, device):
+        pass
+
+    @staticmethod
+    def to_numpy(tensor):
+        if tensor is None:
+            return None
+        return tensor.detach().cpu().numpy() if tensor.requires_grad else tensor.cpu().numpy()
+
+    def stream_step(
```

Comment:
Quite possibly, you can extend ASRModuleMixin with this class and manage CTC/RNNT in one function.

Reply:
ASRModuleMixin is used for the ASR models while this class is for the streaming modules like the encoder of conformer. The stream_step here is is different from the one used for models in the ASRModuleMixin. I already moved the stream_step for models into ASRModuleMixin .

Reply:
Looks like this has been renamed above to conformer_stream_step. Is this function general to multiple models here? If not, please also name it something more unique such as rnnt_stream_step, or chunkwise_stream_step or something.

Reply:
++